From 23846526c8ddd604a3223302734178fa6604c6bc Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Sun, 10 Jan 2016 16:28:15 +0200
Subject: [PATCH 5/7] arm: imx6ul: enable nand for cl-som-imx6ul

cl-som-imx6ul can be equipted with an emmc or nand as
an onboard storage device.
This patch allows recognize what the storage device is;
configures the pads with regards to a connected device.

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 board/compulab/cl_som_imx6ul/cl_som_imx6ul.c | 65 ++++++++++++++++++++++++++++
 include/configs/cl_som_imx6ul.h              | 14 ++++++
 2 files changed, 79 insertions(+)

diff --git a/board/compulab/cl_som_imx6ul/cl_som_imx6ul.c b/board/compulab/cl_som_imx6ul/cl_som_imx6ul.c
index 59df5b5..d8ebd95 100644
--- a/board/compulab/cl_som_imx6ul/cl_som_imx6ul.c
+++ b/board/compulab/cl_som_imx6ul/cl_som_imx6ul.c
@@ -62,6 +62,10 @@ DECLARE_GLOBAL_DATA_PTR;
 	PAD_CTL_SPEED_MED |		\
 	PAD_CTL_DSE_40ohm | PAD_CTL_SRE_FAST)
 
+#ifndef CONFIG_SPL_BUILD
+static int nand_enabled = 0;
+#endif
+
 int dram_init(void)
 {
 	gd->ram_size = imx_ddr_size();
@@ -185,6 +189,11 @@ int board_mmc_init(bd_t *bis)
 			usdhc_cfg[0].sdhc_clk = mxc_get_clock(MXC_ESDHC_CLK);
 			break;
 		case 1:
+			if (nand_enabled) {
+			/* nand enabled configuration */
+				return 0;
+			}
+			/* emmc enabled configuration */
 			imx_iomux_v3_setup_multiple_pads(
 				usdhc2_emmc_pads, ARRAY_SIZE(usdhc2_emmc_pads));
 			usdhc_cfg[1].sdhc_clk = mxc_get_clock(MXC_ESDHC2_CLK);
@@ -424,6 +433,60 @@ static int board_setup_i2c(void)
 static int board_setup_i2c(void) { return 0; }
 #endif
 
+#ifdef CONFIG_NAND_MXS
+static iomux_v3_cfg_t const usdhc2_nand_pads[] = {
+	MX6_PAD_NAND_CLE__RAWNAND_CLE | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_ALE__RAWNAND_ALE | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_WP_B__RAWNAND_WP_B | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_READY_B__RAWNAND_READY_B | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_CE0_B__RAWNAND_CE0_B | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_CE1_B__RAWNAND_CE1_B | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_RE_B__RAWNAND_RE_B | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_WE_B__RAWNAND_WE_B | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA00__RAWNAND_DATA00 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA01__RAWNAND_DATA01 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA02__RAWNAND_DATA02 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA03__RAWNAND_DATA03 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA04__RAWNAND_DATA04 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA05__RAWNAND_DATA05 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA06__RAWNAND_DATA06 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+	MX6_PAD_NAND_DATA07__RAWNAND_DATA07 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+};
+
+#define NAND_ENABLE	IMX_GPIO_NR(4, 13)
+static iomux_v3_cfg_t const nand_enable_pads[] = {
+	MX6_PAD_NAND_CE0_B__GPIO4_IO13 | MUX_PAD_CTRL(USDHC_PAD_CTRL),
+};
+
+static void get_nand_enable_state(void) {
+	imx_iomux_v3_setup_multiple_pads(
+		nand_enable_pads, ARRAY_SIZE(nand_enable_pads));
+	gpio_direction_input(NAND_ENABLE);
+	mdelay(1);
+	nand_enabled = gpio_get_value(NAND_ENABLE);
+}
+
+static void setup_gpmi_nand(void)
+{
+	struct mxc_ccm_reg *mxc_ccm = (struct mxc_ccm_reg *)CCM_BASE_ADDR;
+
+	get_nand_enable_state();
+
+	/* nand enabled configuration */
+	imx_iomux_v3_setup_multiple_pads(
+		usdhc2_nand_pads, ARRAY_SIZE(usdhc2_nand_pads));
+
+	setup_gpmi_io_clk((MXC_CCM_CS2CDR_ENFC_CLK_PODF(0xf) |
+		MXC_CCM_CS2CDR_ENFC_CLK_PRED(1) |
+		MXC_CCM_CS2CDR_ENFC_CLK_SEL(0)));
+
+	/* enable apbh clock gating */
+	setbits_le32(&mxc_ccm->CCGR0, MXC_CCM_CCGR0_APBHDMA_MASK);
+}
+#else
+static void setup_gpmi_nand(void) {}
+#endif
+
 int board_early_init_f(void)
 {
 	setup_iomux_uart();
@@ -436,6 +499,8 @@ int board_init(void)
 	/* Address of boot parameters */
 	gd->bd->bi_boot_params = PHYS_SDRAM + 0x100;
 
+	setup_gpmi_nand();
+
 	setup_fec(CONFIG_FEC_ENET_DEV);
 
 	setup_usb();
diff --git a/include/configs/cl_som_imx6ul.h b/include/configs/cl_som_imx6ul.h
index 0a6ed4f..dbf0beb 100644
--- a/include/configs/cl_som_imx6ul.h
+++ b/include/configs/cl_som_imx6ul.h
@@ -46,6 +46,20 @@
 #define CONFIG_MXC_UART
 #define CONFIG_MXC_UART_BASE		UART3_BASE
 
+/* NAND */
+#ifndef CONFIG_SPL_BUILD
+#define CONFIG_CMD_NAND
+#define CONFIG_SYS_NAND_BASE		0x40000000
+#define CONFIG_SYS_NAND_MAX_CHIPS	1
+#define CONFIG_SYS_MAX_NAND_DEVICE	1
+#define CONFIG_NAND_MXS
+#define CONFIG_SYS_NAND_ONFI_DETECTION
+/* APBH DMA is required for NAND support */
+#define CONFIG_APBH_DMA
+#define CONFIG_APBH_DMA_BURST
+#define CONFIG_APBH_DMA_BURST8
+#endif
+
 /* MMC Configs */
 #define CONFIG_FSL_USDHC
 #ifdef CONFIG_FSL_USDHC
-- 
1.9.1

